File: Blink/Blink/SocketManager.swift
//
//  SocketManager.swift
//  Blink
//
//  Created by Zaki Sheriff on 2025-11-26.
//

import Foundation
import Network
import Combine

// MARK: - Socket Manager
class SocketManager: ObservableObject {
    private var udpListener: NWListener?
    private var tcpListener: NWListener?
    private var connections: [UUID: NWConnection] = [:]
    private let queue = DispatchQueue(label: "com.blink.socket", qos: .userInitiated)
    
    @Published var isListening = false
    @Published var error: String?
    
    // Callbacks
    var onDiscoveryMessage: ((DiscoveryMessage, String) -> Void)?
    var onConnectionReceived: ((NWConnection) -> Void)?
    var onDataReceived: ((Data, UUID) -> Void)?
    
    // MARK: - UDP Broadcast
    func startUDPListener(port: UInt16 = NetworkConstants.discoveryPort) {
        let parameters = NWParameters.udp
        parameters.allowLocalEndpointReuse = true
        
        do {
            udpListener = try NWListener(using: parameters, on: NWEndpoint.Port(rawValue: port)!)
            
            udpListener?.newConnectionHandler = { [weak self] connection in
                self?.handleUDPConnection(connection)
            }
            
            udpListener?.stateUpdateHandler = { [weak self] state in
                DispatchQueue.main.async {
                    switch state {
                    case .ready:
                        self?.isListening = true
                        print("UDP listener ready on port \(port)")
                    case .failed(let error):
                        self?.error = "UDP listener failed: \(error.localizedDescription)"
                        self?.isListening = false
                    default:
                        break
                    }
                }
            }
            
            udpListener?.start(queue: queue)
        } catch {
            DispatchQueue.main.async {
                self.error = "Failed to start UDP listener: \(error.localizedDescription)"
            }
        }
    }
    
    private func handleUDPConnection(_ connection: NWConnection) {
        connection.start(queue: queue)
        
        connection.receiveMessage { [weak self] data, context, isComplete, error in
            if let data = data, !data.isEmpty {
                self?.processUDPMessage(data, from: connection)
            }
            
            // Continue receiving
            if connection.state == .ready {
                self?.handleUDPConnection(connection)
            }
        }
    }
    
    private func processUDPMessage(_ data: Data, from connection: NWConnection) {
        guard let message = try? JSONDecoder().decode(DiscoveryMessage.self, from: data) else {
            return
        }
        
        // Extract IP address from connection
        if case .hostPort(let host, _) = connection.endpoint {
            let ipAddress = "\(host)"
            onDiscoveryMessage?(message, ipAddress)
        }
    }
    
    func sendUDPBroadcast(message: DiscoveryMessage, to address: String, port: UInt16 = NetworkConstants.discoveryPort) {
        guard let data = try? JSONEncoder().encode(message) else { return }
        
        let host = NWEndpoint.Host(address)
        let port = NWEndpoint.Port(rawValue: port)!
        let endpoint = NWEndpoint.hostPort(host: host, port: port)
        
        let connection = NWConnection(to: endpoint, using: .udp)
        connection.start(queue: queue)
        
        connection.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                print("UDP send error: \(error.localizedDescription)")
            }
            connection.cancel()
        })
    }
    
    // MARK: - TCP Server
    func startTCPServer(port: UInt16 = NetworkConstants.transferPort) {
        let parameters = NWParameters.tcp
        parameters.allowLocalEndpointReuse = true
        
        do {
            tcpListener = try NWListener(using: parameters, on: NWEndpoint.Port(rawValue: port)!)
            
            tcpListener?.newConnectionHandler = { [weak self] connection in
                print("New TCP connection received")
                self?.handleTCPConnection(connection)
            }
            
            tcpListener?.stateUpdateHandler = { [weak self] state in
                DispatchQueue.main.async {
                    switch state {
                    case .ready:
                        print("TCP server ready on port \(port)")
                    case .failed(let error):
                        self?.error = "TCP server failed: \(error.localizedDescription)"
                    default:
                        break
                    }
                }
            }
            
            tcpListener?.start(queue: queue)
        } catch {
            DispatchQueue.main.async {
                self.error = "Failed to start TCP server: \(error.localizedDescription)"
            }
        }
    }
    
    private func handleTCPConnection(_ connection: NWConnection) {
        let connectionId = UUID()
        queue.async { [weak self] in
            self?.connections[connectionId] = connection
        }
        
        connection.start(queue: queue)
        onConnectionReceived?(connection)
        
        receiveData(on: connection, id: connectionId)
    }
    
    private func receiveData(on connection: NWConnection, id: UUID) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: NetworkConstants.bufferSize) { [weak self] data, context, isComplete, error in
            if let data = data, !data.isEmpty {
                self?.onDataReceived?(data, id)
            }
            
            if let error = error {
                print("TCP receive error: \(error.localizedDescription)")
                self?.queue.async {
                    self?.connections.removeValue(forKey: id)
                }
                return
            }
            
            if isComplete {
                self?.queue.async {
                    self?.connections.removeValue(forKey: id)
                }
                return
            }
            
            // Continue receiving
            self?.receiveData(on: connection, id: id)
        }
    }
    
    // MARK: - TCP Client
    func connectToDevice(ipAddress: String, port: UInt16, completion: @escaping (NWConnection?, Error?) -> Void) {
        let host = NWEndpoint.Host(ipAddress)
        let port = NWEndpoint.Port(rawValue: port)!
        let endpoint = NWEndpoint.hostPort(host: host, port: port)
        
        let connection = NWConnection(to: endpoint, using: .tcp)
        let connectionId = UUID()
        
        connection.stateUpdateHandler = { [weak self] state in
            switch state {
            case .ready:
                self?.queue.async {
                    self?.connections[connectionId] = connection
                }
                completion(connection, nil)
            case .failed(let error):
                completion(nil, error)
            default:
                break
            }
        }
        
        connection.start(queue: queue)
        receiveData(on: connection, id: connectionId)
    }
    
    func sendData(_ data: Data, on connection: NWConnection, completion: @escaping (Error?) -> Void) {
        connection.send(content: data, completion: .contentProcessed { error in
            completion(error)
        })
    }
    
    // MARK: - Cleanup
    func stop() {
        udpListener?.cancel()
        tcpListener?.cancel()
        
        queue.async { [weak self] in
            guard let self = self else { return }
            for connection in self.connections.values {
                connection.cancel()
            }
            self.connections.removeAll()
        }
        
        DispatchQueue.main.async {
            self.isListening = false
        }
    }
    
    deinit {
        stop()
    }
}

File: android/app/src/main/java/com/example/blink/network/SocketManager.kt
package com.example.blink.network

import android.util.Log
import com.google.gson.Gson
import java.net.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.coroutines.coroutineContext
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class SocketManager {
    private val TAG = "SocketManager"
    private val gson = Gson()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private var udpSocket: DatagramSocket? = null
    private var tcpServerSocket: ServerSocket? = null
    private val connections = ConcurrentHashMap<String, Socket>()

    private val _isListening = MutableStateFlow(false)
    val isListening: StateFlow<Boolean> = _isListening

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    // Callbacks
    var onDiscoveryMessage: ((DiscoveryMessage, String) -> Unit)? = null
    var onConnectionReceived: ((Socket) -> Unit)? = null
    var onDataReceived: ((ByteArray, String) -> Unit)? = null

    // MARK: - UDP Broadcast
    fun startUDPListener(port: Int = NetworkConstants.DISCOVERY_PORT) {
        scope.launch {
            try {
                udpSocket =
                        DatagramSocket(port).apply {
                            broadcast = true
                            reuseAddress = true
                        }

                _isListening.value = true
                Log.d(TAG, "UDP listener started on port $port")

                val buffer = ByteArray(1024)
                while (isActive && udpSocket?.isClosed == false) {
                    try {
                        val packet = DatagramPacket(buffer, buffer.size)
                        udpSocket?.receive(packet)

                        val data = packet.data.copyOf(packet.length)
                        val message = String(data)
                        val ipAddress = packet.address.hostAddress ?: continue

                        try {
                            val discoveryMsg = gson.fromJson(message, DiscoveryMessage::class.java)
                            onDiscoveryMessage?.invoke(discoveryMsg, ipAddress)
                        } catch (e: Exception) {
                            Log.e(TAG, "Failed to parse discovery message", e)
                        }
                    } catch (e: SocketException) {
                        if (isActive) {
                            Log.e(TAG, "UDP receive error", e)
                        }
                        break
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "UDP listener failed", e)
                _error.value = "UDP listener failed: ${e.message}"
                _isListening.value = false
            }
        }
    }

    fun sendUDPBroadcast(
            message: DiscoveryMessage,
            address: String,
            port: Int = NetworkConstants.DISCOVERY_PORT
    ) {
        scope.launch {
            try {
                val json = gson.toJson(message)
                val data = json.toByteArray()

                val socket = DatagramSocket().apply { broadcast = true }

                val packet = DatagramPacket(data, data.size, InetAddress.getByName(address), port)

                socket.send(packet)
                socket.close()
            } catch (e: Exception) {
                Log.e(TAG, "UDP send error", e)
            }
        }
    }

    // MARK: - TCP Server
    fun startTCPServer(port: Int = NetworkConstants.TRANSFER_PORT) {
        scope.launch {
            try {
                tcpServerSocket = ServerSocket(port).apply { reuseAddress = true }

                Log.d(TAG, "TCP server started on port $port")

                while (isActive && tcpServerSocket?.isClosed == false) {
                    try {
                        val socket = tcpServerSocket?.accept() ?: break
                        Log.d(TAG, "New TCP connection from ${socket.inetAddress.hostAddress}")

                        val connectionId = "${socket.inetAddress.hostAddress}:${socket.port}"
                        connections[connectionId] = socket

                        onConnectionReceived?.invoke(socket)

                        // Start receiving data
                        launch { receiveData(socket, connectionId) }
                    } catch (e: SocketException) {
                        if (isActive) {
                            Log.e(TAG, "TCP accept error", e)
                        }
                        break
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "TCP server failed", e)
                _error.value = "TCP server failed: ${e.message}"
            }
        }
    }

    private suspend fun receiveData(socket: Socket, connectionId: String) {
        try {
            val inputStream = socket.getInputStream()
            val buffer = ByteArray(NetworkConstants.BUFFER_SIZE)

            while (coroutineContext.isActive && !socket.isClosed) {
                val bytesRead = inputStream.read(buffer)
                if (bytesRead == -1) break

                val data = buffer.copyOf(bytesRead)
                onDataReceived?.invoke(data, connectionId)
            }
        } catch (e: Exception) {
            Log.e(TAG, "TCP receive error", e)
        } finally {
            connections.remove(connectionId)
            socket.close()
        }
    }

    // MARK: - TCP Client
    suspend fun connectToDevice(ipAddress: String, port: Int): Socket? {
        return withContext(Dispatchers.IO) {
            try {
                val socket = Socket(ipAddress, port)
                val connectionId = "${socket.inetAddress.hostAddress}:${socket.port}"
                connections[connectionId] = socket

                // Start receiving data
                launch { receiveData(socket, connectionId) }

                socket
            } catch (e: Exception) {
                Log.e(TAG, "Failed to connect to device", e)
                null
            }
        }
    }

    suspend fun sendData(data: ByteArray, socket: Socket): Boolean {
        return withContext(Dispatchers.IO) {
            try {
                socket.getOutputStream().write(data)
                socket.getOutputStream().flush()
                true
            } catch (e: Exception) {
                Log.e(TAG, "Failed to send data", e)
                false
            }
        }
    }

    // MARK: - Cleanup
    fun stop() {
        scope.launch {
            udpSocket?.close()
            tcpServerSocket?.close()

            connections.values.forEach { it.close() }
            connections.clear()

            _isListening.value = false
        }
    }

    fun cleanup() {
        stop()
        scope.cancel()
    }
}
